server/src/main/java/org/elasticsearch/search/retriever/RescorerRetrieverBuilder.java

```diff
@@ -0,0 +1,156 @@
+            context.trackRetrieverUsage(innerRetriever.getName());
+            return innerRetriever;
+        }, RETRIEVER_FIELD);
+        PARSER.declareField(constructorArg(), (parser, context) -> {
+            if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
+                List<RescorerBuilder<?>> rescorers = new ArrayList<>();
+                while ((parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                    rescorers.add(RescorerBuilder.parseFromXContent(parser, name -> context.trackRescorerUsage(name)));
+                }
+                return rescorers;
+            } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                return List.of(RescorerBuilder.parseFromXContent(parser, name -> context.trackRescorerUsage(name)));
+            } else {
+                throw new IllegalArgumentException(
+                    "Unknown format for [rescorer.rescore], expects an object or an array of objects, got: " + parser.currentToken()
+                );
+            }
+        }, RESCORE_FIELD, ObjectParser.ValueType.OBJECT_ARRAY);
+        RetrieverBuilder.declareBaseParserFields(NAME, PARSER);
+    }
+
+    public static RescorerRetrieverBuilder fromXContent(XContentParser parser, RetrieverParserContext context) throws IOException {
+        try {
+            return PARSER.apply(parser, context);
+        } catch (Exception e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    private final List<RescorerBuilder<?>> rescorers;
+
+    public RescorerRetrieverBuilder(RetrieverBuilder retriever, List<RescorerBuilder<?>> rescorers) {
+        super(List.of(new RetrieverSource(retriever, null)), extractMinWindowSize(rescorers));
+        if (rescorers.isEmpty()) {
+            throw new IllegalArgumentException("Missing rescore definition");
+        }
+        this.rescorers = rescorers;
+    }
+
+    private RescorerRetrieverBuilder(RetrieverSource retriever, List<RescorerBuilder<?>> rescorers) {
+        super(List.of(retriever), extractMinWindowSize(rescorers));
+        this.rescorers = rescorers;
+    }
+
+    /**
+     * The minimum window size is used as the {@link CompoundRetrieverBuilder#rankWindowSize},
+     * the final number of top documents to return in this retriever.
+     */
+    private static int extractMinWindowSize(List<RescorerBuilder<?>> rescorers) {
+        int windowSize = Integer.MAX_VALUE;
+        for (var rescore : rescorers) {
+            windowSize = Math.min(rescore.windowSize() == null ? RescorerBuilder.DEFAULT_WINDOW_SIZE : rescore.windowSize(), windowSize);
+        }
+        return windowSize;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    protected SearchSourceBuilder finalizeSourceBuilder(SearchSourceBuilder source) {
+        for (var rescorer : rescorers) {
```

Comment:
nit: we could add a comment here on why this differs for non-compound vs compound nested sub-retrievers due to rewrite  (as it is mentioned in the docs).